File: backend/utils/pineconeClient.js
// backend/utils/pineconeClient.js
import { Pinecone } from '@pinecone-database/pinecone';
import dotenv from 'dotenv';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

// Configure dotenv to load environment variables
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
dotenv.config({ path: join(__dirname, '../../.env') });

// Access Pinecone API key and configuration
const apiKey = process.env.PINECONE_API_KEY;
const indexName = process.env.PINECONE_INDEX_NAME || 'fingenie-finance-vectors';
const cloud = process.env.PINECONE_CLOUD || 'aws';
const region = process.env.PINECONE_REGION || 'us-east-1';

// Initialize Pinecone client
let pineconeClient = null;
let pineconeIndex = null;

/**
 * Initialize the Pinecone client and create index if it doesn't exist
 */
export async function initializePinecone() {
  if (!apiKey) {
    console.error('PINECONE_API_KEY is missing in .env file');
    return null;
  }

  try {
    console.log('Initializing Pinecone client...');
    pineconeClient = new Pinecone({
      apiKey: apiKey
    });

    // Check if index exists
    const { indexes } = await pineconeClient.listIndexes();
    const indexExists = indexes.some(idx => idx.name === indexName);

    if (!indexExists) {
      console.log(`Creating new Pinecone index: ${indexName}`);
      await pineconeClient.createIndex({
        name: indexName,
        dimension: 768, // Dimension for embedding-001 model
        metric: 'cosine',
        spec: {
          serverless: {
            cloud: cloud,
            region: region
          }
        }
      });
      
      // Wait for index to be ready
      console.log('Waiting for index to initialize...');
      await new Promise(resolve => setTimeout(resolve, 60000)); // Wait 60 seconds
    }

    // Connect to the index
    pineconeIndex = pineconeClient.index(indexName);
    console.log('Pinecone index connected successfully');
    return pineconeIndex;
  } catch (error) {
    console.error('Error initializing Pinecone:', error);
    return null;
  }
}

/**
 * Get the Pinecone index instance
 */
export function getPineconeIndex() {
  if (!pineconeIndex) {
    console.warn('Pinecone index not initialized, attempting to initialize now...');
    return initializePinecone();
  }
  return pineconeIndex;
}

/**
 * Store embeddings in Pinecone
 * @param {Array} vectors - Array of vectors with id, values, and metadata
 */
export async function storeEmbeddings(vectors) {
  const index = await getPineconeIndex();
  if (!index) {
    console.error('Failed to get Pinecone index');
    return false;
  }

  try {
    // Upsert vectors in batches of 100
    const batchSize = 100;
    for (let i = 0; i < vectors.length; i += batchSize) {
      const batch = vectors.slice(i, i + batchSize);
      await index.upsert(batch);
    }
    console.log(`Successfully stored ${vectors.length} vectors in Pinecone`);
    return true;
  } catch (error) {
    console.error('Error storing embeddings in Pinecone:', error);
    return false;
  }
}

/**
 * Query Pinecone for similar vectors
 * @param {Array} queryEmbedding - The embedding vector to query with
 * @param {Number} topK - Number of results to return
 * @param {Object} filter - Optional metadata filter
 */
export async function querySimilarVectors(queryEmbedding, topK = 5, filter = null) {
  const index = await getPineconeIndex();
  if (!index) {
    console.error('Failed to get Pinecone index');
    return [];
  }

  try {
    // Create query parameters without filter by default
    const queryParams = {
      vector: queryEmbedding,
      topK: topK,
      includeMetadata: true
    };
    
    // Only add filter if it's provided and not empty
    if (filter && Object.keys(filter).length > 0) {
      queryParams.filter = filter;
    }
    
    console.log('Querying Pinecone with params:', JSON.stringify(queryParams));
    const queryResponse = await index.query(queryParams);

    return queryResponse.matches || [];
  } catch (error) {
    console.error('Error querying Pinecone:', error);
    return [];
  }
}

// Initialize Pinecone on module import
initializePinecone().catch(error => {
  console.error('Failed to initialize Pinecone on startup:', error);
});

export default {
  initializePinecone,
  getPineconeIndex,
  storeEmbeddings,
  querySimilarVectors
};

File: fastapi-backend/app/utils/pinecone_client.py
import os
import pinecone
from typing import List, Dict, Any, Optional
import logging
from app.core.config import settings

logger = logging.getLogger(__name__)

# Initialize Pinecone client
def init_pinecone():
    """Initialize the Pinecone client with API key and environment"""
    try:
        pinecone.init(
            api_key=settings.PINECONE_API_KEY,
            environment=settings.PINECONE_CLOUD
        )
        
        # Check if index exists, if not create it
        if settings.PINECONE_INDEX_NAME not in pinecone.list_indexes():
            pinecone.create_index(
                name=settings.PINECONE_INDEX_NAME,
                dimension=768,  # Dimension for Gemini embeddings
                metric="cosine"
            )
            logger.info(f"Created new Pinecone index: {settings.PINECONE_INDEX_NAME}")
        
        # Connect to the index
        index = pinecone.Index(settings.PINECONE_INDEX_NAME)
        logger.info(f"Successfully connected to Pinecone index: {settings.PINECONE_INDEX_NAME}")
        return index
    except Exception as e:
        logger.error(f"Error initializing Pinecone: {str(e)}")
        raise

# Query similar vectors from Pinecone
async def query_similar_vectors(query_embedding: List[float], top_k: int = 5) -> List[Dict[str, Any]]:
    """
    Query Pinecone for vectors similar to the query embedding
    
    Args:
        query_embedding: The embedding vector to query against
        top_k: Number of results to return
        
    Returns:
        List of matches with their scores and metadata
    """
    try:
        # Initialize Pinecone if not already initialized
        index = init_pinecone()
        
        # Query the index
        query_response = index.query(
            vector=query_embedding,
            top_k=top_k,
            include_metadata=True
        )
        
        # Format the response
        matches = []
        for match in query_response.matches:
            matches.append({
                "id": match.id,
                "score": match.score,
                "metadata": match.metadata
            })
        
        return matches
    except Exception as e:
        logger.error(f"Error querying Pinecone: {str(e)}")
        return []

# Upsert vectors to Pinecone
async def upsert_vectors(vectors: List[Dict[str, Any]]) -> Dict[str, Any]:
    """
    Upsert vectors to Pinecone
    
    Args:
        vectors: List of vectors to upsert, each with id, values, and metadata
        
    Returns:
        Response from Pinecone upsert operation
    """
    try:
        # Initialize Pinecone if not already initialized
        index = init_pinecone()
        
        # Upsert the vectors
        response = index.upsert(vectors=vectors)
        
        return response
    except Exception as e:
        logger.error(f"Error upserting vectors to Pinecone: {str(e)}")
        raise

# Delete vectors from Pinecone
async def delete_vectors(ids: List[str]) -> Dict[str, Any]:
    """
    Delete vectors from Pinecone by their IDs
    
    Args:
        ids: List of vector IDs to delete
        
    Returns:
        Response from Pinecone delete operation
    """
    try:
        # Initialize Pinecone if not already initialized
        index = init_pinecone()
        
        # Delete the vectors
        response = index.delete(ids=ids)
        
        return response
    except Exception as e:
        logger.error(f"Error deleting vectors from Pinecone: {str(e)}")
        raise
